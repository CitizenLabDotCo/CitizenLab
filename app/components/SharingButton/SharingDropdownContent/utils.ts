import tracks from '../tracks';
import { trackEventByName } from 'utils/analytics';

export type UtmParams = {
  source: string;
  campaign: string;
  content?: string;
};

export type Medium =
  | 'facebook'
  | 'twitter'
  | 'messenger'
  | 'whatsapp'
  | 'email';

export function addUtmToUrl(
  medium: Medium,
  url: string,
  utmParams?: UtmParams
) {
  let resUrl = url;

  if (utmParams) {
    resUrl += `?utm_source=${utmParams.source}&utm_campaign=${utmParams.campaign}&utm_medium=${medium}`;

    if (utmParams.content) {
      resUrl += `&utm_content=${utmParams.content}`;
    }
  }

  return resUrl;
}

export function handleClick(medium: Medium, href?: string) {
  if (href) {
    window.location.href = href;
  }
  trackEventByName(tracks.clickShare.name, { network: medium });
}

export function trackClick(medium: Medium) {
  trackEventByName(tracks.clickShare.name, { network: medium });
}
