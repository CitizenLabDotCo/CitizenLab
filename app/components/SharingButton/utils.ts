export type UtmParams = {
  source: string;
  campaign: string;
  content?: string;
};

export type Medium =
  | 'facebook'
  | 'twitter'
  | 'messenger'
  | 'whatsapp'
  | 'email';

export function addUtmToUrl(medium: Medium, url: string, utmParams: UtmParams) {
  let resUrl = url;

  resUrl += `?utm_source=${utmParams.source}&utm_campaign=${utmParams.campaign}&utm_medium=${medium}`;

  if (utmParams.content) {
    resUrl += `&utm_content=${utmParams.content}`;
  }

  return resUrl;
}
