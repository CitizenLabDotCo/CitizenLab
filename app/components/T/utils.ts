import { makeSelectSetting } from 'utils/tenant/selectors';
import { makeSelectLocale } from 'containers/LanguageProvider/selectors';
import { Iterable, List } from 'immutable';
import { connect } from 'react-redux';
import { Multiloc, Locale } from 'typings';

const isImmutable = Iterable.isIterable;

const findTranslatedTextMutable = (value: Multiloc | undefined, userLocale: Locale, tenantLocales: List<Locale>): string => {
  let text = '';

  if (!value) {
    return text;
  }

  if (value[userLocale]) {
    text = value[userLocale] || '';
  } else {
    tenantLocales.some((tenantLocale) => {
      if (tenantLocale && value[tenantLocale]) {
        text = value[tenantLocale] || '';
        return true;
      }
      return false;
    });
  }
  return text;
};


function findTranslatedTextImmutable(value: Map<Locale, string>, userLocale: Locale, tenantLocales: List<Locale>): string {
  let text = '';

  if (!value) {
    return text;
  }

  if (value.get(userLocale)) {
    text = value.get(userLocale) || '';
  } else {
    tenantLocales.some((tenantLocale) => {
      if (tenantLocale && value.get(tenantLocale)) {
        text = value.get(tenantLocale) || '';
        return true;
      }
      return false;
    });
  }
  return text;
}

export const findTranslatedText = (value: Multiloc | Map<Locale, string> | undefined, userLocale: Locale, tenantLocales: List<Locale>) => {
  if (!isImmutable(value)) {
    return findTranslatedTextMutable(value as Multiloc, userLocale, tenantLocales);
  }
  return findTranslatedTextImmutable(value as Map<Locale, string>, userLocale, tenantLocales);
};


// Wrap a component with this function to make props.tFunc available. tFunc
// works as the T component, but as a function. This is useful for passing a
// translated value as a prop, e.g. as a placeholder
export function injectTFunc(component) {
  const mapStateToProps = (state) => {
    const selectLocale = makeSelectLocale();
    const selectTenantLocales = makeSelectSetting(['core', 'locales']);
    return {
      tFunc(value) {
        return findTranslatedText(value, selectLocale(state), selectTenantLocales(state));
      },
    };
  };

  return connect(mapStateToProps)(component);
}
