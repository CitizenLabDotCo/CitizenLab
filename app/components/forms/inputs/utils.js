
export const appenDableName = (name) => {
  const camelName = name.replace(/((_|^)\w)/g, (m) => m.slice(-1)[0].toUpperCase());
  return camelName.charAt(0).toUpperCase() + camelName.slice(1);
};

export const toCammelCase = (name) => name.replace(/(_\w)/g, (m) => m.slice(-1)[0].toUpperCase());
