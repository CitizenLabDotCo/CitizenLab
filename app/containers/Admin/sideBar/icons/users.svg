<svg xmlns="http://www.w3.org/2000/svg" width="21" height="22" viewBox="0 0 21 22">
    <g fill="none" fillRule="evenodd" stroke="#FFF" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" opacity=".6" transform="translate(1 1)">
        <circle cx="9.091" cy="10" r="9.091"/>
        <circle cx="9.091" cy="8.977" r="2.537"/>
        <path d="M9.09 12.198a5.074 5.074 0 0 0-5.065 5.346 9.09 9.09 0 0 0 5.066 1.547 9.09 9.09 0 0 0 5.06-1.548c.007-.09.012-.18.014-.27a5.074 5.074 0 0 0-5.074-5.075z"/>
    </g>
</svg>
