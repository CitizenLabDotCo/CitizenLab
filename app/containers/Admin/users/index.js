import React from 'react';
import PropTypes from 'prop-types';
import HelmetIntl from 'components/HelmetIntl';

// components
// import { FormattedMessage } from 'react-intl';
// import WatchSagas from 'utils/containers/watchSagas';
// import sagas from 'resources/users/sagas';

// store
// import { preprocess } from 'utils';

// messages
import messages from './messages';

class UserDashboard extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function

  render() {
    return (
      <div>
        <HelmetIntl
          title={messages.helmetTitle}
          description={messages.helmetDescription}
        />
        {this.props.children}
      </div>

    );
  }
}


UserDashboard.propTypes = {
  children: PropTypes.element,
};

export default UserDashboard;
