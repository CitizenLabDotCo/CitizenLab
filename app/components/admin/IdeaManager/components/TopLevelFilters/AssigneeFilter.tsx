import React, { PureComponent } from 'react';
import { adopt } from 'react-adopt';

import { isNilOrError } from 'utils/helperUtils';

// import Select from 'components/UI/Select';
import { Dropdown } from 'semantic-ui-react';

import GetUsers, { GetUsersChildProps } from 'resources/GetUsers';

import { IOption } from 'typings';

import { injectIntl } from 'utils/cl-intl';
import { InjectedIntlProps } from 'react-intl';
import messages from '../../messages';
import GetAuthUser, { GetAuthUserChildProps } from 'resources/GetAuthUser';

interface DataProps {
  prospectAssignees: GetUsersChildProps;
  authUser: GetAuthUserChildProps;
}

interface InputProps {
  assignee: string;
  handleAssigneeFilterChange: (value: string) => void;
  projectId: string | undefined;
}

interface Props extends InputProps, DataProps {}

interface State {
  assigneeOptions: IOption[];
}

export class AssigneeFilter extends PureComponent<Props & InjectedIntlProps, State> {
  constructor(props) {
    super(props);
    this.state = {
      assigneeOptions: [],
    };
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const { prospectAssignees, authUser, intl: { formatMessage } } = nextProps;
    const nextState = { ...prevState };

    if (prospectAssignees !== prevState.prospectAssignees || authUser !== prevState.authUser) {
      if (isNilOrError(prospectAssignees.usersList) || isNilOrError(authUser)) {
        nextState.assigneeOptions = [];
      } else {
        const assigneeOptionsWithoutCurrentUser = prospectAssignees.usersList.filter(assignee => assignee.id !== authUser.id)
          .map(assignee => ({
            value: assignee.id,
            text: formatMessage(messages.assignedTo, {
              assigneeName:  `${assignee.attributes.first_name} ${assignee.attributes.last_name}`
            })
          }));

        // Order of assignee filter options:
        // All ideas > Assigned to me > Unassigned > Assigned to X (other admins/mods)
        nextState.assigneeOptions = assigneeOptionsWithoutCurrentUser;
        nextState.assigneeOptions.unshift({ value: 'unassigned', text: formatMessage(messages.unassignedIdeas), id: 'e2e-assignee-filter-unassigned' });
        nextState.assigneeOptions.unshift({ value: authUser.id, text: formatMessage(messages.assignedToMe), id: 'e2e-assignee-filter-assigned-to-user' });
        nextState.assigneeOptions.unshift({ value: 'all', text: formatMessage(messages.anyAssignment), id: 'e2e-assignee-filter-all-ideas' });
      }
    }
    return nextState;
  }

  onAssigneeChange = (_event, assigneeOption) => {
    this.props.handleAssigneeFilterChange(assigneeOption.value);
  }

  render() {
    const { assignee } = this.props;
    const { assigneeOptions } = this.state;

    return (
      <Dropdown
        id="e2e-idea-select-assignee-filter"
        options={assigneeOptions}
        onChange={this.onAssigneeChange}
        value={assignee}
        search
        selection
      />
    );
  }
}

const Data = adopt<DataProps, InputProps>({
  prospectAssignees: ({ projectId, render }) => projectId
    ? <GetUsers canModerateProject={projectId} pageSize={250}>{render}</GetUsers>
    : <GetUsers canModerate pageSize={250}>{render}</GetUsers>,
  authUser: <GetAuthUser />
});

const AssigneeFilterWithHocs = injectIntl<Props>(AssigneeFilter);

export default (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <AssigneeFilterWithHocs {...inputProps} {...dataProps} />}
  </Data>
);
