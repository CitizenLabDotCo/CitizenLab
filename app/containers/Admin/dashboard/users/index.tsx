// libraries
import React, { PureComponent } from 'react';
import moment from 'moment';
import { ThemeProvider } from 'styled-components';

// components
import {
  chartTheme,
  Container,
  GraphsContainer,
  Line,
  GraphCard,
  GraphCardInner,
  GraphCardTitle,
  ControlBar
} from '../';
import TimeControl from '../components/TimeControl';
import IntervalControl from '../components/IntervalControl';
import GenderChart from '../components/GenderChart';
import AgeChart from '../components/AgeChart';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from '../messages';

interface State {
  interval: 'weeks' | 'months' | 'years';
  intervalIndex: number;
  currentGroupFilter?: string;
}

type Props = {

};

export default class UsersDashboard extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      interval: 'months',
      intervalIndex: 0,
      currentGroupFilter: undefined
    };
  }

  changeInterval = (interval: 'weeks' | 'months' | 'years') => {
    this.setState({ interval, intervalIndex: 0 });
  }

  changeIntervalIndex = (intervalIndex: number) => {
    this.setState({ intervalIndex });
  }

  handleOnGroupFilter = (filter) => {
    // To be implemented
    this.setState({ currentGroupFilter: filter });
  }

  render() {
    const { interval, intervalIndex, currentGroupFilter } = this.state;
    const startAtMoment = moment().startOf(interval).add(intervalIndex, interval);
    const endAtMoment = moment(startAtMoment).add(1, interval);
    const startAt = startAtMoment.toISOString();
    const endAt = endAtMoment.toISOString();

    return (
      <Container>
        <ControlBar>
          <TimeControl
            value={intervalIndex}
            interval={interval}
            onChange={this.changeIntervalIndex}
            currentTime={startAtMoment}
          />
          <IntervalControl
            value={interval}
            onChange={this.changeInterval}
          />
        </ControlBar>
        <ThemeProvider theme={chartTheme}>
          <GraphsContainer>
            <Line>
              <GraphCard className="first halfWidth">
                <GraphCardInner>
                  <GraphCardTitle>
                    <FormattedMessage {...messages.usersByGenderTitle} />
                  </GraphCardTitle>
                  <GenderChart startAt={startAt} endAt={endAt} currentGroupFilter={currentGroupFilter} />
                </GraphCardInner>
              </GraphCard>
              <GraphCard className="halfWidth">
                <GraphCardInner>
                  <GraphCardTitle>
                    <FormattedMessage {...messages.usersByAgeTitle} />
                  </GraphCardTitle>
                  <AgeChart startAt={startAt} endAt={endAt} currentGroupFilter={currentGroupFilter} />
                </GraphCardInner>
              </GraphCard>
            </Line>
          </GraphsContainer>
        </ThemeProvider>
      </Container>
    );
  }
}
