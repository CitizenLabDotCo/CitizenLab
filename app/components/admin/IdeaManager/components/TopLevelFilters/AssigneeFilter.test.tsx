import React from 'react';
import { shallow } from 'enzyme';

import { AssigneeFilter } from './AssigneeFilter';
import { makeUser } from 'services/__mocks__/users';

jest.mock('utils/cl-intl');
jest.mock('resources/GetUsers');
jest.mock('services/users');
jest.mock('services/auth');
import { intl } from 'utils/cl-intl';

describe('<AssigneeFilter />', () => {
  let handleAssigneeFilterChange: jest.Mock;
  beforeEach(() => {
    handleAssigneeFilterChange = jest.fn();
  });

  it('processes options correctly', () => {
    const authUser = makeUser({ roles: [{ type: 'admin' }] }, 'me').data;
    const prospectAssignees = ['admin1', 'admin2', 'admin3'].map(name => makeUser({ slug: name, first_name: name, roles: [{ type: 'admin' }] }, name).data);
    prospectAssignees.push(authUser);

    const wrapper = shallow(
      <AssigneeFilter
        intl={intl}
        authUser={authUser}
        handleAssigneeFilterChange={handleAssigneeFilterChange}
        prospectAssignees={{ usersList: prospectAssignees }}
        assignee="me"
        projectId={undefined}
      />
    );
    expect(wrapper.find('Dropdown').prop('options')).toMatchSnapshot();
  });

  it('passes the user ID to handleAssigneeOnChange', () => {
    const authUser = makeUser({ roles: [{ type: 'admin' }] }, 'me').data;
    const prospectAssignees = ['admin1', 'admin2', 'admin3'].map(name => makeUser({ slug: name, first_name: name, roles: [{ type: 'admin' }] }, name).data);
    prospectAssignees.push(authUser);

    const wrapper = shallow(
      <AssigneeFilter
        intl={intl}
        authUser={authUser}
        handleAssigneeFilterChange={handleAssigneeFilterChange}
        prospectAssignees={{ usersList: prospectAssignees }}
        assignee="me"
        projectId={undefined}
      />
    );
    const options = wrapper.find('Dropdown').prop('options');
    const mine = options.find(option => option.value === 'me');
    wrapper.instance().onAssigneeChange({}, mine);
    expect(handleAssigneeFilterChange.mock.calls[0][0]).toBe('me');
  });

  it('passes down keyword all if you select All ideas', () => {
    const authUser = makeUser({ roles: [{ type: 'admin' }] }, 'me').data;
    const prospectAssignees = ['admin1', 'admin2', 'admin3'].map(name => makeUser({ slug: name, first_name: name, roles: [{ type: 'admin' }] }, name).data);
    prospectAssignees.push(authUser);

    const wrapper = shallow(
      <AssigneeFilter
        intl={intl}
        authUser={authUser}
        handleAssigneeFilterChange={handleAssigneeFilterChange}
        prospectAssignees={{ usersList: prospectAssignees }}
        assignee="me"
        projectId={undefined}
      />
    );
    const options = wrapper.find('Dropdown').prop('options');
    const all = options.find(option => option.text === 'All ideas');
    wrapper.instance().onAssigneeChange({}, all);
    expect(handleAssigneeFilterChange.mock.calls[0][0]).toBe('all');
  });
});
