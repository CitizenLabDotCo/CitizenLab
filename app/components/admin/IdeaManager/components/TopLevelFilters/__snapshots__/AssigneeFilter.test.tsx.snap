// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssigneeFilter /> processes options correctly 1`] = `
Array [
  Object {
    "text": "All ideas",
    "value": "all",
  },
  Object {
    "text": "Assigned to admin1 Citizenlab",
    "value": "admin1",
  },
  Object {
    "text": "Assigned to admin2 Citizenlab",
    "value": "admin2",
  },
  Object {
    "text": "Assigned to admin3 Citizenlab",
    "value": "admin3",
  },
  Object {
    "text": "Assigned to me",
    "value": "me",
  },
]
`;
